label_studio/ml/models.py

```diff
@@ -258,6 +258,70 @@ def _predict(self, task):
             },
         }
 
+    def _get_predictions_from_ml_backend(self, serialized_tasks: List[Dict]) -> List[Dict]:
+        result = self.api.make_predictions(serialized_tasks, self.project)
+
+        # response validation
+        if result.is_error:
+            logger.error(f'Error occurred: {result.error_message}')
+            return []
+        elif not isinstance(result.response, dict) or 'results' not in result.response:
+            logger.error(f'ML backend returns an incorrect response, it must be a dict: {result.response}')
+            return []
+        elif not isinstance(result.response['results'], list) or len(result.response['results']) == 0:
+            logger.error('ML backend returns an incorrect response, it must be a list with at least one result')
+            return []
+
+        responses = result.response['results']
+
+        predictions = []
+        if len(serialized_tasks) != len(responses):
+            if len(responses) == 1:
+                # In case ML backend doesn't support batch of tasks, do it one by one
+                # TODO: remove this block after all ML backends will support batch processing
+                logger.warning(
+                    f"'ML backend '{self.title}' doesn't support batch processing of tasks, "
+                    f'switched to one-by-one task retrieval'
+                )
+                for serialized_task in serialized_tasks:
+                    # get predictions per task
+                    predictions.extend(self._get_predictions_from_ml_backend([serialized_task]))
+
+                return predictions
+            else:
+                # complete failure - likely ML backend skipped some tasks, we can't match them
+                logger.error(
+                    f'Number of tasks and responses are not equal: '
+                    f'{len(serialized_tasks)} tasks != {len(responses)} responses. '
+                    f'Returning empty predictions.'
+                )
+                return []
```

Comment:
This function is a bit long - could we maybe factor out this part I've highlighted to a helper method?

Reply:
fixed